src/lib/storage/index/group_key/group_key_index.cpp

```diff
@@ -14,41 +15,67 @@ size_t GroupKeyIndex::estimate_memory_consumption(ChunkOffset row_count, ChunkOf
+  Assert(static_cast<bool>(_indexed_segment), "GroupKeyIndex only works with dictionary segments_to_index.");
   Assert((segments_to_index.size() == 1), "GroupKeyIndex only works with a single segment.");
 
-  // 1) Initialize the index structures
-  // 1a) Set the index_offset to size of the dictionary + 1 (plus one to mark the ending position)
-  //     and set all offsets to 0
-  _index_offsets = std::vector<size_t>(_indexed_segments->unique_values_count() + 1u, 0u);
-  // 1b) Set the _index_postings to the size of the attribute vector
-  _index_postings = std::vector<ChunkOffset>(_indexed_segments->size());
+  // 1) Initialize the index_offset:
+  //    Set the index_offset to size of the dictionary + 1 (plus one to mark the ending position)
+  //    and set all offsets to 0.
+  //    Using `_index_offsets`, we want to count the occurences of each ValueID of the attribute vector (AV).
+  //    The ValueID for NULL in an AV is the highest available ValueID in the dictionary + 1.
+  //    `unique_values_count` returns the size of dictionary which does not store a ValueID for NULL.
+  //    Therefore we have `unique_values_count` ValueIDs (null value id is not included)
+  //    for which we want to count the occurrences.
+  _index_offsets =
+      std::vector<ChunkOffset>(_indexed_segment->unique_values_count() + 1u /*to mark the ending position */, 0u);
 
   // 2) Count the occurrences of value-ids: Iterate once over the attribute vector (i.e. value ids)
   //    and count the occurrences of each value id at their respective position in the dictionary,
   //    i.e. the position in the _index_offsets
-  resolve_compressed_vector_type(*_indexed_segments->attribute_vector(), [&](auto& attribute_vector) {
+  const auto null_value_id = _indexed_segment->null_value_id();
+  auto null_count = 0u;
+
+  resolve_compressed_vector_type(*_indexed_segment->attribute_vector(), [&](auto& attribute_vector) {
     for (const auto value_id : attribute_vector) {
-      _index_offsets[value_id + 1u]++;
+      if (static_cast<ValueID>(value_id) != null_value_id) {
+        _index_offsets[value_id + 1u]++;
+      } else {
+        ++null_count;
+      }
     }
   });
 
-  // 3) Create offsets for the postings in _index_offsets
+  const auto non_null_count = _indexed_segment->size() - null_count;
+
+  // 3) Set the _index_postings and _index_null_positions
+  _index_postings = std::vector<ChunkOffset>(non_null_count);
+  _index_null_positions = std::vector<ChunkOffset>(null_count);
+
+  // 4) Create offsets for the postings in _index_offsets
   std::partial_sum(_index_offsets.begin(), _index_offsets.end(), _index_offsets.begin());
 
-  // 4) Create the postings
-  // 4a) Copy _index_offsets to use it as a write counter
-  auto index_offset_copy = std::vector<size_t>(_index_offsets);
+  // 5) Create the postings
+  // 5a) Copy _index_offsets to use it as a write counter
```

Comment:
Why? I don't see _index_offsets being modified after this. What do we need the copy for?

Reply:
_index_offsets is not modified after this but the copy is.
We increment the offset of the ValueID for which position was added to _index_postings in the corresponding iteration (see Snippet 1).
Since _index_offsets is used in `_get_postings_iterator_at(ValueID value_id) const` (see Snippet 2), modifying a ValueID offset in the original _index_offsets instead of in the copy would lead to an incorrect return value of `_get_postings_iterator_at(ValueID value_id) const`.

Snippet 1:
https://github.com/hyrise/hyrise/blob/7143a3fea58cb05134f5e2806d64b0bfbb1063fd/src/lib/storage/index/group_key/group_key_index.cpp#L80-L82

Snippet 2:
https://github.com/hyrise/hyrise/blob/7143a3fea58cb05134f5e2806d64b0bfbb1063fd/src/lib/storage/index/group_key/group_key_index.cpp#L114-L126

Renaming might clarify this:
`index_offset_coy` -> `index_insert_offsets`
`_index_offsets` -> `_index_start_offsets` (not absolutely necessary but increases the informative value)

Reply:
Sounds good.